Notify framework to clear input connection when app is backgrounded (#35054).

matthew-carroll opened the pull request:
Problem:
Flutter is not setup to handle backgrounding and then foregrounding an app when an input control has focus. The issue is visible in the new embedding because the new embedding loosens platform restrictions around `FlutterActivity` continuity.

Repro Steps:

- Run an app with a TextField
- Give focus to the TextField
- Press Android's home button
- Switch back to the app
- Tap the TextField to give it focus again
- An error appears in the logs.

The issue is that Flutter is unaware that the app was backgrounded and therefore expects the same input connection to exist when re-foregrounded. But on the Android side, the input connection was lost when the app was backgrounded and therefore needs to be recreated when foregrounded again.

Solution:
When `FlutterActivity` goes to the background and detaches from the `FlutterEngine` instance, instruct the framework to clear the input client. By clearing it, when the user taps on the control after being re-foregrounded, Flutter will request a new input connection and things should continue as expected.

Corresponding Framework PR:
https://github.com/flutter/flutter/pull/35100

xster commented:
I assume this also fixes the case of a user opening a flutter activity, focuses on text, backs out of that flutter activity while keeping the same engine, then relaunches the same activity with the same engine and focuses on text again (assuming the same problem would exist there too)?

amirh commented on shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:
```diff
@@ -127,6 +127,9 @@ public void unlockPlatformViewInputConnection() {
      * The TextInputPlugin instance should not be used after calling this.
      */
     public void destroy() {
+        if (inputTarget.type != InputTarget.Type.NO_TARGET) {
```
We should only do this when the target is a FRAMEWORK_CLIENT (if the target is a PLATFORM_VIEW `_currentConnection` on the framework side is probably null).

matthew-carroll commented:
I updated the PR with a different approach that utilizes `InputConnection#closeConnection()` instead of `FlutterView` destruction. This change in approach now solves the original but, as well as a bug where giving focus to a regular Android View did not result in Flutter UI from losing focus, or clearing of Flutter's input connection.

I also renamed the Android side message from 'clearClient' to 'onConnectionClosed'.

I'm still dubious about where and how I'm invoking `unfocus()`. If anyone can see a more canonical place to adjust focus, I'd like to hear it.

xster commented on shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java:
```diff
@@ -222,6 +223,17 @@ public void unspecifiedAction(int inputClientId) {
+   * Instructs Flutter to clear the current input client, which ends the text
+   * input interaction with the given input control.
+   */
+  public void onConnectionClosed(int inputClientId) {
```
nit: if the rest of the file are verbs, let's make this a verb too (e.g. closeConnection)

matthew-carroll replied:
I originally had it as a verb but then changed it. This one is phrased as an event because it has already occurred. In this scenario, Flutter is not closing anything, Android already closed the input connection and then informed Flutter that it had closed. Does that make sense?

xster replied:
ah sure. SG

xster commented on shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:
```diff
@@ -127,6 +134,9 @@ public void unlockPlatformViewInputConnection() {
      * The TextInputPlugin instance should not be used after calling this.
      */
     public void destroy() {
+//        if (inputTarget.type != InputTarget.Type.NO_TARGET) {
```
what's this for?

matthew-carroll replied:
That's an oversight. I'll fix it.

mklim commented on shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java:
```diff
@@ -36,20 +34,33 @@ public InputConnectionAdaptor(
+
+        textInputChannel.onConnectionClosed(mClient);
+
+        if (onConnectionClosed != null) {
```
nit: This is final and never null, so I think it's okay to just always run this. Not a big deal and it doesn't hurt to be safe though.

related: annotations may be worthwhile in the constructor?

matthew-carroll replied:
Good point, I'll add annotations.

matthew-carroll replied:
The parameters are now annotated and the if-statement has been removed.

matthew-carroll commented:
Rebasing to resolve conflicts.

cbracken commented:
@matthew-carroll is this mergeable?

matthew-carroll commented:
I have to figure out why the tests are failing and fix them.

matthew-carroll commented:
@xster @mklim I added a unit test

mklim commented on testing/embedding/android/focus_new_activity_back/.flutter-plugins:
```diff
@@ -0,0 +1 @@
+android_intent=/Users/mattcarroll/.pub-cache/hosted/pub.dartlang.org/android_intent-0.3.1+1/
```
Pretty sure this should be reverted. We probably should add files with this extension to the `.gitignore`.

xster commented:
+1. It's a bit hard to review since all the build intermediaries are checked in.

xster commented:
LGTM
